scheduler/src/cook/mesos/api.clj

```diff
@@ -57,15 +83,98 @@
+;;   :env "MYENV abc"
+;;   :container { :type "docker"
+;;     :docker {:image "bp4_res_20150904.patched"
+;;              :parameters [ {:env "KRB5CCNAME=${KRB5CCNAME}" } {:user "10138"}]
+;;              :network "HOST"
+;;             }
+;;   :volumes [ {:containerPath "/tmp/$KRB5CCNAME} :hostPath "/tmp/$KRB5CCNAME" :mode "RO"}]
+;;
+;; }
+;;
+;;
+;;
+
+;; helper for submit-jobs, deal with container structure
+(defn build-container [id container]
+  (let [container-id (d/tempid :db.part/user)
+        docker-id (d/tempid :db.part/user)
+        ctype (:type container)
+        volumes (if (contains? container :volumes)
+                  (:volumes container)
+                  [])
+        mkparams (fn [cid params]
+                   (let [param-id (d/tempid :db.part/user)]
+                         (if (empty? params) {}
+                             {:docker/parameters
+                              (mapv (fn [kvp]
+                                      (let
+                                          [k (:key kvp)
+                                           v (:value kvp)
+                                           ]
+                                        {:docker.param/key k
+                                         :docker.param/value v}
+                                        )) params)})))
+        mkvolumes (fn [cid vols]
+                    (let [modemap { "RO" 2 "RW" 1 }]
+                      (if (empty? vols) {}
+                          {:container/volumes
+                           (mapv
+                            (fn [m]
+                              (reduce
+                               (fn [b kvp]
+                                 (into b
+                                       (cond
+                                         (= (key kvp) :container_path)
+                                         {:container.volume/container_path (val kvp)}
+                                         (= (key kvp) :host_path)
+                                         {:container.volume/host_path (val kvp)}
+                                         (= (key kvp) :mode)
+                                         {:container.volume/mode ;;(get modemap (clojure.string/upper-case (val kvp)))}
+                                          (val kvp)}
+                                         ))) {} m))vols)})))]
+    (cond
+      (= (clojure.string/lower-case ctype) "docker")
+      (let [docker (:docker container)
+            params (if (contains? docker :parameters)
+                     (:parameters docker)
+                     [])
+            netmap { "HOST" 1 "BRIDGE" 2 "NONE" 3}]
+        [[:db/add id :job/container container-id]
+         (into {:db/id container-id
+                :container/type "DOCKER"}
+               ;;:container/type 1 }
+               (mkvolumes container-id volumes))
+         [:db/add container-id :container/docker docker-id]
+         (into {:db/id docker-id
```

Comment:
this should be part of the `mkparams` (which should probably have a different name)